cli/src/commands/parse/msgs.rs

```diff
@@ -0,0 +1,426 @@
+    loop {
+        let attachment_path = get_attachment_store_path(attachment_number);
+
+        if compound_file.is_storage(attachment_path.clone()) {
+            attachments.push(read_attachment(attachment_path, &mut compound_file)?);
+        } else {
+            break;
+        }
+
+        attachment_number += 1;
+    }
+
+    // User Properties
+    let mut user_properties = PropertyMap::new();
+    user_properties.insert_string(
+        MSG_NAME_USER_PROPERTY_NAME.to_string(),
+        path.file_name()
+            .context("Could not get file name")?
+            .to_string_lossy()
+            .to_string(),
+    );
+
+    Ok(Document {
+        raw_email: RawEmail {
+            body: RawEmailBody::Plain(plain_body_string),
+            headers: RawEmailHeaders::Raw(headers_string_no_content_headers),
+            attachments,
+        },
+        user_properties,
+        comment_id: None,
+    })
+}
+
+fn upload_batch_of_documents(
+    client: &Client,
+    source: &Source,
+    documents: &Vec<Document>,
+    transform_tag: &TransformTag,
+    no_charge: bool,
+    statistics: &Arc<Statistics>,
+) -> Result<()> {
+    client.sync_raw_emails(
+        &source.full_name(),
+        documents,
+        transform_tag,
+        false,
+        no_charge,
+    )?;
+    statistics.add_uploaded(documents.len());
+    Ok(())
+}
+
+pub fn get_msgs_in_directory(directory: &PathBuf) -> Result<Vec<DirEntry>> {
+    Ok(std::fs::read_dir(directory)?
+        .filter_map(|path| {
+            if let Ok(path) = path {
+                if !path.path().extension().is_some_and(|msg| msg == "msg") {
+                    None
+                } else {
+                    Some(path)
+                }
+            } else {
+                None
+            }
+        })
+        .collect())
+}
+
+pub fn parse(client: &Client, args: &ParseMsgArgs) -> Result<()> {
+    let ParseMsgArgs {
+        directory,
+        source,
+        transform_tag,
+        no_charge,
+        yes,
+    } = args;
+
+    if !no_charge && !yes {
+        ensure_uip_user_consents_to_ai_unit_charge(client.base_url())?;
+    }
+
+    let msg_paths = get_msgs_in_directory(directory)?;
+    let statistics = Arc::new(Statistics::new());
+    let _progress = get_progress_bar(msg_paths.len() as u64, &statistics);
+    let source = client.get_source(source.clone())?;
+
+    let mut documents = Vec::new();
+    let mut errors = Vec::new();
+    for path in msg_paths {
+        match read_msg_to_document(&path.path()) {
+            Ok(document) => {
+                documents.push(document);
+
+                if documents.len() >= UPLOAD_BATCH_SIZE {
+                    upload_batch_of_documents(
+                        client,
+                        &source,
+                        &documents,
+                        transform_tag,
+                        *no_charge,
+                        &statistics,
+                    )?;
+                    documents.clear();
```

Comment:
As you call this more than once (again after the loop), you can easily pull this out into a closure:

```rust
    let send = |documents: &mut Vec<Document>| -> Result<()> {
        upload_batch_of_documents(
            client,
            &source,
            documents,
            transform_tag,
            *no_charge,
            &statistics,
        )?;
        documents.clear();
        Ok(())
    };
```